value/src/main/java/com/google/auto/value/AutoValueExtension.java

```diff
@@ -0,0 +1,33 @@
+package com.google.auto.value;
+
+import javax.annotation.processing.ProcessingEnvironment;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.TypeElement;
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * Created by rharter on 5/1/15.
+ */
+public interface AutoValueExtension {
+
+  public interface Context {
+    ProcessingEnvironment processingEnvironment();
+    String packageName();
+    TypeElement autoValueClass();
+    Map<String, ExecutableElement> properties();
+  }
+
+  public interface GeneratedClass {
+    String className();
+    String source();
+    Collection<String> additionalImports();
+    Collection<ExecutableElement> consumedProperties();
+    Collection<String> additionalInterfaces();
+    Collection<String> additionalCode();
```

Comment:
Well yes, but that would mean AutoValueProcessor would need to parse the code output by extensions to see whether it defines toString(). That can be done (by deferring the generation of the final AutoValue implementation to the next round of annotation processing) but it's not simple. I'm also not keen on a solution that is a mix of generating one class per extension and generating further code to be injected into the original implementation class.

Incidentally another issue with the JavaPoet approach is that extensions would need to apply the same Maven-shading as AutoValue so that their references to JavaPoet APIs would be match the APIs that AutoValue would be using. In other words, if AutoValue renames com.squareup.javapoet.TypeName to autovalue.com.squareup.javapoet.TypeName, then extensions will either have to reference that renamed class in their source code, or apply the same renaming through Maven shading when packaged.